File: Beecrowd/5 - Mathematics/1163 - Angry Ducks/Python/1163_AngryDucks.py
import math
import sys

PI = 3.14159
GRAVITY = 9.80665

input = sys.stdin.read
data = input().split()

index = 0

while index < len(data):
    initialHeight = float(data[index])
    index += 1
    minDistance = float(data[index])
    maxDistance = float(data[index + 1])
    numShots = int(data[index + 2])
    index += 3

    for _ in range(numShots):
        angleDegrees = float(data[index])
        initialSpeed = float(data[index + 1])
        index += 2

        angleRadians = angleDegrees * PI / 180

        horizontalSpeed = initialSpeed * math.cos(angleRadians)
        verticalSpeed = initialSpeed * math.sin(angleRadians)

        timeToRise = verticalSpeed / GRAVITY

        maxHeight = (verticalSpeed * verticalSpeed) / (2 * GRAVITY) + initialHeight

        finalVerticalSpeed = math.sqrt(2 * GRAVITY * maxHeight)

        timeToFall = finalVerticalSpeed / GRAVITY

        totalTime = timeToRise + timeToFall

        horizontalDistance = horizontalSpeed * totalTime

        if minDistance < horizontalDistance < maxDistance:
            print(f"{horizontalDistance:.5f} -> DUCK")
        else:
            print(f"{horizontalDistance:.5f} -> NUCK")

File: Beecrowd/5 - Mathematics/1163 - Angry Ducks/1163_AngryDucks.cpp
#include <bits/stdc++.h>

using namespace std;

#define PI 3.14159
#define GRAVITY 9.80665

int main() {
    double initialHeight;

    while (cin >> initialHeight) {
        int numShots;
        double minDistance, maxDistance;
        cin >> minDistance >> maxDistance >> numShots;

        for (int i = 0; i < numShots; i++) {
            double angleDegrees, initialSpeed;
            cin >> angleDegrees >> initialSpeed;

            double angleRadians = angleDegrees * PI / 180;

            double horizontalSpeed, verticalSpeed;
            horizontalSpeed = initialSpeed * cos(angleRadians);
            verticalSpeed = initialSpeed * sin(angleRadians);

            double timeToRise;
            timeToRise = verticalSpeed / GRAVITY;

            double maxHeight;
            maxHeight =
                (verticalSpeed * verticalSpeed) / (2 * GRAVITY) + initialHeight;

            double finalVerticalSpeed;
            finalVerticalSpeed = sqrt(2 * GRAVITY * maxHeight);

            double timeToFall;
            timeToFall = finalVerticalSpeed / GRAVITY;

            double totalTime;
            totalTime = timeToRise + timeToFall;

            double horizontalDistance;
            horizontalDistance = horizontalSpeed * totalTime;

            if (horizontalDistance > minDistance &&
                horizontalDistance < maxDistance)
                cout << fixed << setprecision(5) << horizontalDistance
                     << " -> DUCK" << endl;
            else
                cout << fixed << setprecision(5) << horizontalDistance
                     << " -> NUCK" << endl;
        }
    }

    return 0;
}
